subsys/testsuite/ztest/src/ztress.c

```diff
@@ -0,0 +1,421 @@
+ */
+static void adjust_load(uint8_t ratio)
+{
+	for (uint32_t i = 0; i < context_cnt; i++) {
+		uint32_t new_ticks = ratio * (uint32_t)backoff[i].ticks / 16;
+
+		backoff[i].ticks = MAX(4, new_ticks);
+	}
+}
+
+static void progress_timeout(struct k_timer *timer)
+{
+	struct ztress_context_data *thread_data = k_timer_user_data_get(timer);
+	uint32_t progress = 100;
+	uint32_t cnt = context_cnt;
+
+	if (tmr_data) {
+		cnt--;
+		if (tmr_data->exec_cnt && exec_cnt[cnt]) {
+			progress = (100 * exec_cnt[cnt]) / tmr_data->exec_cnt;
+		}
+	}
+
+	for (uint32_t i = 0; i < cnt; i++) {
+		if (!thread_data[i].exec_cnt && !thread_data[i].preempt_cnt) {
+			continue;
+		}
+
+		uint32_t exec_progress = (thread_data[i].exec_cnt) ?
+				(100 * exec_cnt[i]) / thread_data[i].exec_cnt : 100;
+		uint32_t preempt_progress = (thread_data[i].preempt_cnt) ?
+				(100 * preempt_cnt[i]) / thread_data[i].preempt_cnt : 100;
+		uint32_t thread_progress = MIN(exec_progress, preempt_progress);
+
+		progress = MIN(progress, thread_progress);
+	}
+
+
+	uint64_t rem = 1000 * (k_timer_expires_ticks(&test_timer) - sys_clock_tick_get()) /
+			CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+
+	printk("\r%u%% remaining:%u ms", progress, (uint32_t)rem);
+}
+
+static void control_load(void)
+{
+	static uint64_t prev_cycles;
+	static uint64_t total_cycles;
+
+	k_thread_runtime_stats_t rt_stats_thread;
+	k_thread_runtime_stats_t rt_stats_all;
+	int err = 0;
+
+	err = k_thread_runtime_stats_get(idle_tid, &rt_stats_thread);
+	if (err < 0) {
+		return;
+	}
+
+	err = k_thread_runtime_stats_all_get(&rt_stats_all);
+	if (err < 0) {
+		return;
+	}
+
+	int load = 1000 - (1000 * (rt_stats_thread.execution_cycles - prev_cycles) /
+			(rt_stats_all.execution_cycles - total_cycles));
+
+	prev_cycles = rt_stats_thread.execution_cycles;
+	total_cycles = rt_stats_all.execution_cycles;
+
+	int avg_load = (rt.cpu_load * rt.cpu_load_measurments + load) /
+			(rt.cpu_load_measurments + 1);
+
+	rt.cpu_load = avg_load;
+	rt.cpu_load_measurments++;
+
+	if (load > 800 && load < 850) {
+		/* Expected load */
+	} else if (load > 850) {
+		/* Slightly reduce load. */
+		adjust_load(18);
+	} else if (load < 300) {
+		adjust_load(8);
+	} else if (load < 500) {
+		adjust_load(12);
+	} else {
+		adjust_load(14);
+	}
+}
+
+static void ctrl_timeout(struct k_timer *timer)
+{
+	control_load();
+}
+
+void preempt_update(void)
+{
+	uint32_t mask = active_mask;
+
+	while (mask) {
+		int idx = 31 - __builtin_clz(mask);
+
+		/* Clear mask to ensure that other context does not count same thread. */
+		if (atomic_and(&active_mask, ~BIT(idx)) & BIT(idx)) {
+			preempt_cnt[idx]++;
+		}
+
+		mask &= ~BIT(idx);
+	}
+}
+
+static bool cont_check(struct ztress_context_data *context_data, uint32_t priority)
+{
+	if (context_data->preempt_cnt && preempt_cnt[priority] >= context_data->preempt_cnt) {
+		atomic_dec(&active_cnt);
+		return false;
+	}
+
+	if (context_data->exec_cnt && exec_cnt[priority] >= context_data->exec_cnt) {
```

Comment:
Again, as per my understanding of the coding guidelines, both context_data->preempt_cnt and context_data->exec_cnt should be explicitly compared against not being zero (!= 0). (As this is the 2nd time that I mentioned the coding guidelines, I am not going to mention it again in this review should there be another instance.)